File: WD.HabitTracker.Infra.Persistence.Mongodb/Repositories/UserRepository.cs
using MongoDB.Driver;
using TWD.HabitTracker.Application.Infra.Persistence.Habits;
using TWD.HabitTracker.Application.Infra.Persistence.Users;
using TWD.HabitTracker.Domain.Entities.Habits;
using TWD.HabitTracker.Domain.Entities.User;
using WD.HabitTracker.Infra.Persistence.Mongodb.Documents;
using WD.HabitTracker.Infra.Persistence.Mongodb.Documents.Users;

namespace WD.HabitTracker.Infra.Persistence.Mongodb.Repositories;

public class UserRepository : IUserReadRepository, IUserWriteRepository
{
    private readonly IMongoCollection<UserDocument> _userCollection;
    
    public UserRepository(HabitTrackerMongoDatabaseSettings settings)
    {
        var client = new MongoClient(settings.ConnectionString);
        var database = client.GetDatabase(settings.DatabaseName);

        _userCollection = database.GetCollection<UserDocument>(settings.UsersCollectionName);
    }

    public async Task AddAsync(User user)
    {
        var doc = new UserDocument(user);
        await _userCollection.InsertOneAsync(doc);
    }

    public async Task<User?> FindByDeviceTokenAsync(string deviceToken)
    {
        var userDocQuery = await _userCollection.FindAsync(d => d.AuthInfo.DeviceAuth != null && d.AuthInfo.DeviceAuth.DeviceToken == deviceToken);
        var userDoc = await userDocQuery.FirstOrDefaultAsync();

        return userDoc?.ToUser();
    }
}

File: WD.HabitTracker.Persistence.MongoDb/UserRepository.fs
namespace WD.HabitTracker.Persistence.MongoDb

open Microsoft.FSharp.Control
open MongoDB.Driver
open WD.HabitTracker.Application.Services.Persistence
open WD.HabitTracker.Domain.Users
open WD.HabitTracker.Persistence.MongoDb.Documents

type UserRepository(client: HabitTrackerMongoDatabase) =
    let userCollection = client.GetUserCollection ()
    
    let mapUserDocumentToUSer userDocument =
        {
            Id = userDocument.Id
            AuthenticationMean = Device {
                DeviceToken = userDocument.AuthInfo.DeviceAuth.DeviceToken
            }
        }
    
    interface IUserReadRepository with
        member this.FindUserByDeviceTokenAsync(deviceToken) = async {

            let filter = Builders<UserDocument>.Filter.Eq((fun u -> u.AuthInfo.DeviceAuth.DeviceToken), deviceToken)
            
            let! matchingUsers = userCollection.FindAsync<UserDocument>(filter) |> Async.AwaitTask
            
            return match (matchingUsers.ToList() |> Seq.tryHead) with
                    | None -> None
                    | Some userDocument -> userDocument |> mapUserDocumentToUSer |> Some
        }